### ISSUE TITLE
Maybe/Result without the 'Value' property (enforce checking before use)

### ISSUE THREAD

**Issue by Szune** [vkhorikov/CSharpFunctionalExtensions#356]
First off - great work on the project, it's been a very helpful tool in C#!

I was unable to find a discussion on this topic, so if you've already had one[0] and this is simply rehashing stuff, I wouldn't mind an instant closing of the issue :)

The one thing I'm noticing that I run into sometimes when handling a Maybe/Result is that I forget to check "HasValue" in cases where you can't (or just don't know yet how to) naturally fit it all neatly into a method chain, leading to an exception being thrown.

I made my own project to handle that single case[1] for both Maybe and Result but later realized it is a little silly to do that instead of checking if I can contribute to your project :)

So, instead of me trying to get my project on par with yours, I'd like to hear what you think of any of:
a) A new CSharpFunctionalExtensions project specifically without .Value and obsolete methods, which I could help maintain instead of maintaining my own project
b) A roslyn analyzer that warns about uses of .Value without a prior checking of .HasValue or .IsSuccess
c) Same project, new backwards incompatible version
d) Same project, new types, keeping backwards compatibility
e) Your suggestion here (can't imagine I've thought of everything)
f) None of the above

Sadly there are of course cons to all my proposed "solutions".
The new project would mean they won't be compatible with each other, causing a bit of a fractured mess for existing projects.
The new types could cause confusion about which to use for new users.

I've no idea of the performance impact and complexity the roslyn analyzer would mean, but I imagine as long as you limit the check to the current and the previous scope, it wouldn't be too bad.
But then it could miss stuff and might lead to a false sense of security, which isn't great either.

---
[0] this is hardly groundbreaking, I imagine it is likely you've already thought of all of this :)
[1] in essence just trying to make sure that it is "practically" impossible to make the mistake of using .Value or .Error before making sure there is something there, to try to eliminate as many exceptions as possible.
In my particular case I went with .ValueOr() (this project's .GetValueOrDefault(), I prefer your name though), .Map() and .IsSome(out T)
Out variables aren't perfect and there are scoping issues with them as well, but without the .Value property I find it a little bit clearer that I have to be aware of the lack of a value. It is of course highly subjective :)

**Comment by vkhorikov** [comment]
We had a similar discussion here: https://github.com/vkhorikov/CSharpFunctionalExtensions/issues/312

Ideally, a roslyn analyzer would be best here (point b), but I'm not sure about the maintenance impact it might entail for the whole library. Another option is to mark `Value` as obsolete, but I don't quite like this solution.

The simplest approach is to encourage the use of `GetValueOrThrow()` and `GetValueOrDefault()` over `Value` in XML comments and in documentation.

Backward incompatibility is not an option, unfortunately.

**Comment by Szune** [comment]
I agree, that does seem to make the most sense when trying to cause the least amount of disruption.
I think I'll have a look at writing a roslyn analyzer prototype to see what it would mean in terms of maintenance and performance.
If it works well enough, would you prefer I put it in my own repository (to not burden you with requests for maintenance), or that I make a pull request here?

I have caught the mistake previously thanks to the XML comment on Value as well, so that's good :)

**Comment by vkhorikov** [comment]
Let's try it as a separate nuget package, at least initially. I'd rather to keep the scope of this library as narrow as possible.

**Comment by AlmarAubel** [comment]
Something likes this? @vkhorikov  https://github.com/AlmarAubel/CSharpFunctionalExtensions.Analyzers? I recently started working on a analyser package for this library. Currently focussed on "Accessing Value without checking IsSuccess or IsError can result in an error"

**Comment by vkhorikov** [comment]
Yes! Feel free to update the readme with a link to your project and submit a PR. This set of analyzers would be useful for a lot of people using this library.
